### DISCUSSION TITLE
握手时`IHAVE`消息中chunkhash数量的问题

### DISCUSSION THREAD

**Discussion by ZephyrusZhang** [SUSTech-CS305-Fall22/CS305-Project-Skeleton#24] [Q&A] [answered]
请问在Section 2.3 中提到在握手时，peer在收到`WHOHAS`包时，peer会回复一个`IHAVE`包，并在data中表明当前peer有的chunk的哈希值。请问这里即便peer有多个被请求的chunk，他也是只回复某一个他有的chunk吗？（因为文档中用的是 "peer sends IHAVE packet containing the **hash** of the requested chunks that it owns"）。如果是只回复一个的话，在peer有多个的情况下应该回复哪一个呢？

**Comment by TANSixu** [comment, marked as answer]
hi, 非常抱歉，这里应该是 **hashes**

如果有多个被请求的，他应该回复所有自己有的hash。但是一次GET只包含一个hash。
例如，A啥也没有，B有1，2，3，C有4，5，

A收到下载1，2，4的命令：

* A->B,C: WHOHAS 1,2,4?
* B->A: IHAVE 1,2
* C->A: IHAVE 4
* A->B: GET 1
* A->C: GET 4
* concurrently download 1,4
* A->B: GET 2
* download 2 from B

**Reply by ZephyrusZhang** [reply]
如果A请求的chunk，B、C都有的话，这种情况又是怎么处理的呢？

**Reply by TANSixu** [reply]
> 如果A请求的chunk，B、C都有的话，这种情况又是怎么处理的呢？

可以都记录下来，如果在后续传输过程中其中一个peer掉线了，那么可以直接找另一个peer下载。
在一开始选择找谁下载的时候不需要实现真实p2p的选择策略，可以认为选择策略不会影响测试结果。
* 在comprehensive测试中，只需要完成下载即可
* 在bonus中，只会测试congestion control，所有peer下载的内容都不同，不需要考虑找更近的peer下载。

**Comment by YanWQ-monad** [comment]
顺便想问一下，是否需要考虑 hashes 数量非常多的情况。因为如果包大小最大是 1400 的话，每个 hash 的大小最少也要 20 bytes，那样的话只能装 70 个左右，那样的话，是否需要考虑超过 70 个的情况呢？~~或者更直接地，测试会不会覆盖到这个地步~~

**Reply by TANSixu** [reply]
hi，不需要考虑，测试的下载量都不大的～
